### PR TITLE
net: pkt_filter: Introduce additional hooks for pkt_filter 

### PR DESCRIPTION
**Pull request by MarGasiorek**
The additional hooks provide infrastructure to construct rules on another network stack levels.
Main benefit of this approach is packets are pre-parsed and e.g. IP filter is easier to implement.
These hooks are equivalent of prerouting and local_in in linux's netfilter.

### CODE REVIEW

File: include/zephyr/net/net_pkt.h [hunk 1]
```diff
@@ -1275,6 +1275,37 @@ static inline bool net_pkt_filter_recv_ok(struct net_pkt *pkt)
...
+static inline bool net_pkt_filter_ip_recv_ok(struct net_pkt *pkt)
+{
+	ARG_UNUSED(pkt);
+
+	return true;
+}
+
+#endif /* CONFIG_NET_PKT_FILTER_IPV4_HOOK || CONFIG_NET_PKT_FILTER_IPV6_HOOK */
+
+#if defined(CONFIG_NET_PKT_FILTER) && defined(NET_PKT_FILTER_LOCAL_IN_HOOK)
```
Comment (rlubos) [inline include/zephyr/net/net_pkt.h:1294]:
```suggestion
#if defined(CONFIG_NET_PKT_FILTER) && defined(CONFIG_NET_PKT_FILTER_LOCAL_IN_HOOK)
```

File: include/zephyr/net/net_pkt.h [hunk 2]
```diff
@@ -1275,6 +1275,37 @@ static inline bool net_pkt_filter_recv_ok(struct net_pkt *pkt)
...
+#else
+
+static inline bool net_pkt_filter_local_in_recv_ok(struct net_pkt *pkt)
+{
+	ARG_UNUSED(pkt);
+
+	return true;
+}
+
+#endif /* CONFIG_NET_PKT_FILTER && NET_PKT_FILTER_LOCAL_IN_HOOK */
```
Comment (rlubos) [inline include/zephyr/net/net_pkt.h:1307]:
```suggestion
#endif /* CONFIG_NET_PKT_FILTER && CONFIG_NET_PKT_FILTER_LOCAL_IN_HOOK */
```

File: include/zephyr/net/net_pkt_filter.h [hunk 3]
```diff
@@ -111,6 +117,27 @@ bool npf_remove_all_rules(struct npf_rule_list *rules);
 #define npf_remove_all_send_rules() npf_remove_all_rules(&npf_send_rules)
 #define npf_remove_all_recv_rules() npf_remove_all_rules(&npf_recv_rules)
 
+#ifdef NET_PKT_FILTER_LOCAL_IN_HOOK
```
Comment (rlubos) [inline include/zephyr/net/net_pkt_filter.h:120]:
```suggestion
#ifdef CONFIG_NET_PKT_FILTER_LOCAL_IN_HOOK
```

File: include/zephyr/net/net_pkt_filter.h [hunk 4]
```diff
@@ -111,6 +117,27 @@ bool npf_remove_all_rules(struct npf_rule_list *rules);
...
+#define npf_insert_local_in_recv_rule(rule) npf_insert_rule(&npf_local_in_recv_rules, rule)
+#define npf_append_local_in_recv_rule(rule) npf_append_rule(&npf_local_in_recv_rules, rule)
+#define npf_remove_local_in_recv_rule(rule) npf_remove_rule(&npf_local_in_recv_rules, rule)
+#define npf_remove_all_local_in_recv_rules() npf_remove_all_rules(&npf_local_in_recv_rules)
+#endif /* NET_PKT_FILTER_LOCAL_IN_HOOK */
```
Comment (rlubos) [inline include/zephyr/net/net_pkt_filter.h:125]:
```suggestion
#endif /* CONFIG_NET_PKT_FILTER_LOCAL_IN_HOOK */
```

File: subsys/net/pkt_filter/base.c [hunk 5]
```diff
@@ -25,6 +25,50 @@ struct npf_rule_list npf_recv_rules = {
 	.lock = { },
 };
 
+#ifdef NET_PKT_FILTER_LOCAL_IN_HOOK
```
Comment (rlubos) [inline subsys/net/pkt_filter/base.c:28]:
```suggestion
#ifdef CONFIG_NET_PKT_FILTER_LOCAL_IN_HOOK
```

File: subsys/net/pkt_filter/base.c [hunk 6]
```diff
@@ -98,6 +142,31 @@ bool net_pkt_filter_recv_ok(struct net_pkt *pkt)
 	return result == NET_OK;
 }
 
+#ifdef NET_PKT_FILTER_LOCAL_IN_HOOK
```
Comment (rlubos) [inline subsys/net/pkt_filter/base.c:145]:
```suggestion
#ifdef CONFIG_NET_PKT_FILTER_LOCAL_IN_HOOK
```

File: subsys/net/pkt_filter/base.c [hunk 7]
```diff
@@ -98,6 +142,31 @@ bool net_pkt_filter_recv_ok(struct net_pkt *pkt)
...
+bool net_pkt_filter_local_in_recv_ok(struct net_pkt *pkt)
+{
+	enum net_verdict result = lock_evaluate(&npf_local_in_recv_rules, pkt);
+
+	return result == NET_OK;
+}
+#endif /* NET_PKT_FILTER_LOCAL_IN_HOOK */
```
Comment (rlubos) [inline subsys/net/pkt_filter/base.c:152]:
```suggestion
#endif /* CONFIG_NET_PKT_FILTER_LOCAL_IN_HOOK */
```

**Review by rlubos — COMMENTED** [review]
There are a few `CONFIG_` prefixes missing (maybe we should also add a simple test for that hook?) but otherwise LGTM.

**Review by rlubos — APPROVED** [review]

**Review by npitre — APPROVED** [review]

File: include/zephyr/net/net_pkt_filter.h [hunk 8]
```diff
@@ -296,6 +323,60 @@ extern npf_test_fn_t npf_size_inbounds;
...
+ * This tests if the packet source ip address matches any of the ip
+ * addresses contained in the provided set.
+ *
+ * @param _name Name of the condition
+ * @param _ip_addr_array Array of <tt>struct in_addr</tt> or <tt>struct in6_addr</tt> items to test
+ *against
+ * @param _ip_addr_num number of IP addresses in the array
+ * @param _af Addresses family type (AF_INET / AF_INET6) in the array
+ */
+#define NPF_IP_SRC_ADDR_ALLOWLIST(_name, _ip_addr_array, _ip_addr_num, _af) \
```
Comment (jukkar) [inline include/zephyr/net/net_pkt_filter.h:352]: I am just wondering that shouldn't we have network interface (either index value or the pointer) here as a parameter? I am thinking about use case where there are multiple network interfaces and we want to block certain addresses either in one specific network interface or all of them.

Comment (MarGasiorek) [reply include/zephyr/net/net_pkt_filter.h:352]: @jukkar I thought about something like you mentioned, but as a additional rule, e.g. 
```suggestion
#define NPF_IP_SRC_ADDR_ALLOWLIST_4_IFC(_name, _ip_addr_array, _ip_addr_num, _af) \
```
Additionally a user can will make by himself custom rule based on the structure related to this macro.
We can add it now or in future, what is your opinion?

Comment (carlescufi) [reply include/zephyr/net/net_pkt_filter.h:352]: Oops, missed this discussion before merging. @MarGasiorek and @jukkar sorry, and feel free to continue the discussion and follow-up with a PR.

Comment (jukkar) [reply include/zephyr/net/net_pkt_filter.h:352]: No worries, additional features can be added in future PRs.

**Pull request merged**